File: DogHandoffCard/HandoffReadiness.swift
import Foundation

enum HandoffReadinessSeverity: String {
    case blocker
    case warning
    case info
}

struct HandoffReadinessItem: Identifiable, Hashable {
    let id: String
    let title: String
    let detail: String
    let severity: HandoffReadinessSeverity
}

struct HandoffReadinessReport: Hashable {
    let score: Int
    let items: [HandoffReadinessItem]

    var blockers: [HandoffReadinessItem] {
        items.filter { $0.severity == .blocker }
    }

    var warnings: [HandoffReadinessItem] {
        items.filter { $0.severity == .warning }
    }

    var hasBlockers: Bool {
        !blockers.isEmpty
    }

    var statusTitle: String {
        if hasBlockers {
            return "Needs critical details"
        }
        if score >= 90 {
            return "Ready to share"
        }
        if score >= 75 {
            return "Usable with notes"
        }
        return "Needs review"
    }

    var statusDescription: String {
        if hasBlockers {
            return "Add the critical items before sending this handoff card."
        }
        if warnings.isEmpty {
            return "The card includes the core details a temporary caregiver needs."
        }
        return "The card can be shared, but a few optional details would make it safer."
    }

    var priorityItems: [HandoffReadinessItem] {
        items.sorted { lhs, rhs in
            severityRank(lhs.severity) < severityRank(rhs.severity)
        }
    }

    private func severityRank(_ severity: HandoffReadinessSeverity) -> Int {
        switch severity {
        case .blocker:
            return 0
        case .warning:
            return 1
        case .info:
            return 2
        }
    }
}

enum HandoffReadiness {
    static func evaluate(for pet: PetProfile, draft: HandoffCardDraft? = nil) -> HandoffReadinessReport {
        var items: [HandoffReadinessItem] = []

        if pet.name.isBlank {
            items.append(
                .init(
                    id: "missing-dog-name",
                    title: "Add the dog name",
                    detail: "The caregiver needs a clear dog name on the exported card.",
                    severity: .blocker
                )
            )
        }

        if pet.ageText.isBlank && pet.breed.isBlank {
            items.append(
                .init(
                    id: "missing-profile-context",
                    title: "Add age or breed",
                    detail: "Age or breed helps caregivers identify the dog and understand senior-care context.",
                    severity: .warning
                )
            )
        }

        evaluateCareRules(pet: pet, items: &items)
        evaluateMedications(pet: pet, items: &items)
        evaluateContacts(pet: pet, items: &items)

        if let draft = draft {
            evaluateDraft(draft, items: &items)
        }

        let blockerPenalty = items.filter { $0.severity == .blocker }.count * 18
        let warningPenalty = items.filter { $0.severity == .warning }.count * 8
        let score = max(0, min(100, 100 - blockerPenalty - warningPenalty))

        return HandoffReadinessReport(score: score, items: items)
    }

    private static func evaluateCareRules(pet: PetProfile, items: inout [HandoffReadinessItem]) {
        guard let careRule = pet.primaryCareRule else {
            items.append(
                .init(
                    id: "missing-care-rules",
                    title: "Add daily care rules",
                    detail: "Feeding, water, walks, and no-go rules are the core of a temporary handoff.",
                    severity: .blocker
                )
            )
            return
        }

        if careRule.feedingAmount.isBlank {
            items.append(
                .init(
                    id: "missing-feeding",
                    title: "Add feeding instructions",
                    detail: "A caregiver should not need to guess meal amount or frequency.",
                    severity: .blocker
                )
            )
        }

        if careRule.walkFrequency.isBlank {
            items.append(
                .init(
                    id: "missing-walks",
                    title: "Add walk frequency",
                    detail: "Walk timing keeps the day predictable for a senior dog.",
                    severity: .warning
                )
            )
        }

        if careRule.callOwnerNowText.isBlank {
            items.append(
                .init(
                    id: "missing-call-owner",
                    title: "Add call-owner triggers",
                    detail: "List the signs that should make the caregiver call you immediately.",
                    severity: .warning
                )
            )
        }

        if careRule.goVetNowText.isBlank {
            items.append(
                .init(
                    id: "missing-vet-guidance",
                    title: "Add emergency vet guidance",
                    detail: "The handoff should say when to go directly to a vet.",
                    severity: .warning
                )
            )
        }
    }

    private static func evaluateMedications(pet: PetProfile, items: inout [HandoffReadinessItem]) {
        if pet.sortedMedications.isEmpty {
            items.append(
                .init(
                    id: "no-medications",
                    title: "Confirm medication status",
                    detail: "If the dog takes no medication, leave this as a conscious empty state.",
                    severity: .info
                )
            )
            return
        }

        for medication in pet.sortedMedications {
            let displayName = medication.name.ifBlank("Unnamed medication")

            if medication.name.isBlank {
                items.append(
                    .init(
                        id: "missing-medication-name-\(medication.primarySortKey)",
                        title: "Name every medication",
                        detail: "One medication entry is missing its name.",
                        severity: .blocker
                    )
                )
            }

            if medication.dosage.isBlank {
                items.append(
                    .init(
                        id: "missing-dose-\(displayName)",
                        title: "Add dose for \(displayName)",
                        detail: "Dose must be explicit before a temporary caregiver gives medication.",
                        severity: .blocker
                    )
                )
            }

            if medication.timeSlots.isEmpty && medication.frequencyText.isBlank {
                items.append(
                    .init(
                        id: "missing-schedule-\(displayName)",
                        title: "Add timing for \(displayName)",
                        detail: "Use exact times or a clear frequency such as twice daily.",
                        severity: .blocker
                    )
                )
            }

            if medication.missedDoseInstruction.isBlank {
                items.append(
                    .init(
                        id: "missing-missed-dose-\(displayName)",
                        title: "Add missed-dose rule for \(displayName)",
                        detail: "This avoids double-dosing or unsafe make-up doses.",
                        severity: .warning
                    )
                )
            }
        }
    }

    private static func evaluateContacts(pet: PetProfile, items: inout [HandoffReadinessItem]) {
        let owner = pet.sortedContacts.first { $0.role == "Owner" }
        if owner == nil || owner?.phone.isBlank == true {
            items.append(
                .init(
                    id: "missing-owner-phone",
                    title: "Add owner phone",
                    detail: "The exported card needs an owner contact that works outside the app.",
                    severity: .blocker
                )
            )
        }

        let clinic = pet.sortedContacts.first { $0.role == "Clinic" }
        if clinic == nil || clinic?.phone.isBlank == true {
            items.append(
                .init(
                    id: "missing-clinic-phone",
                    title: "Add clinic phone",
                    detail: "A vet or clinic phone should be available in emergencies.",
                    severity: .warning
                )
            )
        }
    }

    private static func evaluateDraft(_ draft: HandoffCardDraft, items: inout [HandoffReadinessItem]) {
        if draft.caregiverName.isBlank {
            items.append(
                .init(
                    id: "missing-caregiver-name",
                    title: "Add caregiver name",
                    detail: "Naming the recipient makes reused cards less ambiguous.",
                    severity: .warning
                )
            )
        }

        if draft.endDate < draft.startDate {
            items.append(
                .init(
                    id: "invalid-coverage-window",
                    title: "Fix coverage dates",
                    detail: "End time should be after the start time.",
                    severity: .blocker
                )
            )
        }
    }
}

private extension String {
    var isBlank: Bool {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}

File: app/src/main/java/com/example/doghandoffcard/HandoffReadiness.java
package com.example.doghandoffcard;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public final class HandoffReadiness {
    private HandoffReadiness() {
    }

    static HandoffReadinessReport evaluate(DogProfile profile, HandoffDraft draft) {
        List<HandoffReadinessItem> items = new ArrayList<>();

        if (isBlank(profile.name)) {
            items.add(HandoffReadinessItem.blocker(
                "missing-dog-name",
                "Add the dog name",
                "The caregiver needs a clear dog name on the card."
            ));
        }

        if (isBlank(profile.age) && isBlank(profile.breed)) {
            items.add(HandoffReadinessItem.warning(
                "missing-profile-context",
                "Add age or breed",
                "Age or breed helps caregivers identify the dog and understand senior-care context."
            ));
        }

        evaluateCareRules(profile.careRule, items);
        evaluateMedications(profile.sortedMedications(), items);
        evaluateContacts(profile.sortedContacts(), items);
        evaluateDraft(draft, items);

        int blockers = 0;
        int warnings = 0;
        for (HandoffReadinessItem item : items) {
            if (item.severity == HandoffReadinessSeverity.BLOCKER) {
                blockers++;
            } else if (item.severity == HandoffReadinessSeverity.WARNING) {
                warnings++;
            }
        }

        int score = Math.max(0, Math.min(100, 100 - blockers * 18 - warnings * 8));
        return new HandoffReadinessReport(score, items);
    }

    private static void evaluateCareRules(CareRule rule, List<HandoffReadinessItem> items) {
        if (isBlank(rule.feeding)) {
            items.add(HandoffReadinessItem.blocker(
                "missing-feeding",
                "Add feeding instructions",
                "A caregiver should not need to guess meal amount or frequency."
            ));
        }

        if (isBlank(rule.walks)) {
            items.add(HandoffReadinessItem.warning(
                "missing-walks",
                "Add walk frequency",
                "Walk timing keeps the day predictable for a senior dog."
            ));
        }

        if (isBlank(rule.callOwnerIf)) {
            items.add(HandoffReadinessItem.warning(
                "missing-call-owner",
                "Add call-owner triggers",
                "List the signs that should make the caregiver call you immediately."
            ));
        }

        if (isBlank(rule.goVetIf)) {
            items.add(HandoffReadinessItem.warning(
                "missing-vet-guidance",
                "Add emergency vet guidance",
                "The handoff should say when to go directly to a vet."
            ));
        }
    }

    private static void evaluateMedications(List<MedicationItem> medications, List<HandoffReadinessItem> items) {
        if (medications.isEmpty()) {
            items.add(HandoffReadinessItem.info(
                "no-medications",
                "Confirm medication status",
                "If the dog takes no medication, leave this as a conscious empty state."
            ));
            return;
        }

        for (MedicationItem medication : medications) {
            String displayName = fallback(medication.name, "Unnamed medication");

            if (isBlank(medication.name)) {
                items.add(HandoffReadinessItem.blocker(
                    "missing-medication-name",
                    "Name every medication",
                    "One medication entry is missing its name."
                ));
            }

            if (isBlank(medication.dosage)) {
                items.add(HandoffReadinessItem.blocker(
                    "missing-dose-" + safeId(displayName),
                    "Add dose for " + displayName,
                    "Dose must be explicit before a temporary caregiver gives medication."
                ));
            }

            if (isBlank(medication.schedule)) {
                items.add(HandoffReadinessItem.blocker(
                    "missing-schedule-" + safeId(displayName),
                    "Add timing for " + displayName,
                    "Use exact times or a clear frequency such as twice daily."
                ));
            }

            if (isBlank(medication.missedDoseRule)) {
                items.add(HandoffReadinessItem.warning(
                    "missing-missed-dose-" + safeId(displayName),
                    "Add missed-dose rule for " + displayName,
                    "This avoids double-dosing or unsafe make-up doses."
                ));
            }
        }
    }

    private static void evaluateContacts(List<EmergencyContact> contacts, List<HandoffReadinessItem> items) {
        EmergencyContact owner = findContact(contacts, "Owner");
        if (owner == null || isBlank(owner.phone)) {
            items.add(HandoffReadinessItem.blocker(
                "missing-owner-phone",
                "Add owner phone",
                "The exported card needs an owner contact that works outside the app."
            ));
        }

        EmergencyContact clinic = findContact(contacts, "Clinic");
        if (clinic == null || isBlank(clinic.phone)) {
            items.add(HandoffReadinessItem.warning(
                "missing-clinic-phone",
                "Add clinic phone",
                "A vet or clinic phone should be available in emergencies."
            ));
        }
    }

    private static void evaluateDraft(HandoffDraft draft, List<HandoffReadinessItem> items) {
        if (isBlank(draft.caregiverName)) {
            items.add(HandoffReadinessItem.warning(
                "missing-caregiver-name",
                "Add caregiver name",
                "Naming the recipient makes reused cards less ambiguous."
            ));
        }

        if (draft.endAt.isBefore(draft.startAt)) {
            items.add(HandoffReadinessItem.blocker(
                "invalid-coverage-window",
                "Fix coverage dates",
                "End time should be after the start time."
            ));
        }
    }

    private static EmergencyContact findContact(List<EmergencyContact> contacts, String role) {
        for (EmergencyContact contact : contacts) {
            if (role.equals(contact.role)) {
                return contact;
            }
        }
        return null;
    }

    static boolean isBlank(String value) {
        return value == null || value.trim().isEmpty();
    }

    static String fallback(String value, String fallback) {
        return isBlank(value) ? fallback : value.trim();
    }

    private static String safeId(String value) {
        return value.toLowerCase().replaceAll("[^a-z0-9]+", "-").replaceAll("(^-|-$)", "");
    }
}

enum HandoffReadinessSeverity {
    BLOCKER,
    WARNING,
    INFO
}

final class HandoffReadinessItem {
    final String id;
    final String title;
    final String detail;
    final HandoffReadinessSeverity severity;

    private HandoffReadinessItem(String id, String title, String detail, HandoffReadinessSeverity severity) {
        this.id = id;
        this.title = title;
        this.detail = detail;
        this.severity = severity;
    }

    static HandoffReadinessItem blocker(String id, String title, String detail) {
        return new HandoffReadinessItem(id, title, detail, HandoffReadinessSeverity.BLOCKER);
    }

    static HandoffReadinessItem warning(String id, String title, String detail) {
        return new HandoffReadinessItem(id, title, detail, HandoffReadinessSeverity.WARNING);
    }

    static HandoffReadinessItem info(String id, String title, String detail) {
        return new HandoffReadinessItem(id, title, detail, HandoffReadinessSeverity.INFO);
    }
}

final class HandoffReadinessReport {
    final int score;
    final List<HandoffReadinessItem> items;

    HandoffReadinessReport(int score, List<HandoffReadinessItem> items) {
        this.score = score;
        this.items = new ArrayList<>(items);
    }

    boolean hasBlockers() {
        for (HandoffReadinessItem item : items) {
            if (item.severity == HandoffReadinessSeverity.BLOCKER) {
                return true;
            }
        }
        return false;
    }

    String statusTitle() {
        if (hasBlockers()) {
            return "Needs critical details";
        }
        if (score >= 90) {
            return "Ready to share";
        }
        if (score >= 75) {
            return "Usable with notes";
        }
        return "Needs review";
    }

    String statusDescription() {
        if (hasBlockers()) {
            return "Add the critical items before sending this handoff card.";
        }
        for (HandoffReadinessItem item : items) {
            if (item.severity == HandoffReadinessSeverity.WARNING) {
                return "The card can be shared, but a few optional details would make it safer.";
            }
        }
        return "The card includes the core details a temporary caregiver needs.";
    }

    List<HandoffReadinessItem> priorityItems() {
        List<HandoffReadinessItem> sorted = new ArrayList<>(items);
        sorted.sort(Comparator.comparingInt(item -> severityRank(item.severity)));
        return sorted;
    }

    private int severityRank(HandoffReadinessSeverity severity) {
        switch (severity) {
            case BLOCKER:
                return 0;
            case WARNING:
                return 1;
            case INFO:
            default:
                return 2;
        }
    }
}
